catroid/src/main/java/org/catrobat/catroid/transfers/project/ProjectDownloadService.kt

```diff
@@ -0,0 +1,165 @@
+import org.catrobat.catroid.R
+import org.catrobat.catroid.common.Constants
+import org.catrobat.catroid.common.Constants.CACHE_DIR
+import org.catrobat.catroid.common.Constants.CATROBAT_EXTENSION
+import org.catrobat.catroid.common.Constants.MAX_PERCENT
+import org.catrobat.catroid.common.Constants.TMP_DIR_NAME
+import org.catrobat.catroid.common.FlavoredConstants
+import org.catrobat.catroid.io.XstreamSerializer
+import org.catrobat.catroid.io.ZipArchiver
+import org.catrobat.catroid.utils.ToastUtil
+import org.catrobat.catroid.utils.notifications.StatusBarNotificationManager
+import org.catrobat.catroid.web.CatrobatServerCalls
+import org.catrobat.catroid.web.CatrobatWebClient
+import java.io.File
+import java.io.IOException
+
+class ProjectDownloadService : IntentService("ProjectDownloadService") {
+
+    companion object {
+        val TAG: String = ProjectDownloadService::class.java.simpleName
+        const val EXTRA_DOWNLOAD_NAME = "downloadName"
+        const val EXTRA_URL = "url"
+        const val EXTRA_RENAME_AFTER_DOWNLOAD = "renameAfterDownload"
+        const val EXTRA_RESULT_RECEIVER = "receiver"
+        const val EXTRA_NOTIFICATION_ID = "id"
+        private const val DOWNLOAD_FILE_NAME = "down$CATROBAT_EXTENSION"
+
+        const val UPDATE_PROGRESS_EXTRA = "progress"
+
+        const val UPDATE_PROGRESS_CODE = 1
+        const val ERROR_CODE = 2
+        const val SUCCESSFUL_CODE = 3
+    }
+
+    override fun onHandleIntent(intent: Intent?) {
+        val downloadIntent = intent
+            ?: return logWarning("Called ProjectDownloadService with null intent - aborting")
+        val projectName = downloadIntent.getStringExtra(EXTRA_DOWNLOAD_NAME)
+            ?: return logWarning("Called ProjectDownloadService with null projectName -  aborting")
+        val url = downloadIntent.getStringExtra(EXTRA_URL)
+            ?: return logWarning("Called ProjectDownloadService without url - aborting")
+        val renameAfterDownload = downloadIntent.getBooleanExtra(EXTRA_RENAME_AFTER_DOWNLOAD, false)
+        val resultReceiver = downloadIntent.getParcelableExtra<ResultReceiver>(EXTRA_RESULT_RECEIVER)
+                ?: return logWarning("Called ProjectDownloadService without url - aborting")
+
+        val zipFileString = File(File(CACHE_DIR, TMP_DIR_NAME), DOWNLOAD_FILE_NAME).absolutePath
+        val destinationFile = File(zipFileString)
+
+        if ((destinationFile.parentFile.isDirectory or destinationFile.parentFile.mkdirs()).not()) {
+            ToastUtil.showError(this, R.string.error_project_download)
+            return
+        }
+
+        val id = downloadIntent.getIntExtra(EXTRA_NOTIFICATION_ID, 0)
+        val statusBarNotificationManager = StatusBarNotificationManager.getInstance()
+        val notification = statusBarNotificationManager.getNotification(this, id)
+
+        // startForeground does't do the same thing in previous versions of Android
+        // (it makes the notification buggy)
```

Comment:
Exact same problem. StartForeground attaches the notification to the Service and StopForeground (false) should detach it, but unfortunatly it doesn't (The notification gets dismissed after the execution of the service). Luckily stopForeground(STOP_FOREGROUND_DETACH) works fine, but is only available after version O. Since we don't need the StartForeground method before version O...  I'll try to write better comments ^^ Or should I refactor it to your version?

Reply:
I'd prefer to have as few switches regarding the current platform version as possible. So if `ProjectUploadService` is working across platforms, I'd rather have a similar solution to that (removing the id with `stopForeground(true)`, then issuing a new notification with a different id in success/error callbacks).